File: android/sdk/src/main/java/com/viant/forgeandroid/ui/ReportBuilderAuthoredRuntime.kt
package com.viant.forgeandroid.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.size
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.viant.forgeandroid.runtime.ChartAxisDef
import com.viant.forgeandroid.runtime.ChartDef
import com.viant.forgeandroid.runtime.ChartSeriesDef
import com.viant.forgeandroid.runtime.ChartValueOption
import com.viant.forgeandroid.runtime.ContainerDef
import com.viant.forgeandroid.runtime.ControlState
import com.viant.forgeandroid.runtime.DashboardReportBuilderDef
import com.viant.forgeandroid.runtime.ForgeRuntime
import com.viant.forgeandroid.runtime.InlineReportRuntimeCompiler
import com.viant.forgeandroid.runtime.JsonUtil
import com.viant.forgeandroid.runtime.ReportBuilderPublishedDataSourceDef
import com.viant.forgeandroid.runtime.WindowContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.coroutines.flow.first

internal fun reportBuilderAuthoredDocument(windowForm: Map<String, Any?>): JsonObject? {
    fun obj(value: Any?): JsonObject? = JsonUtil.anyToElement(value) as? JsonObject
    val reportDefinition = obj(windowForm["reportDefinition"])
    val definitionDocument = obj(reportDefinition?.get("documentPatch"))
        ?: obj(reportDefinition?.get("reportDocument"))
    // Conversation window snapshots depth-limit deeply nested reportDefinition
    // values. The report-builder state carries the same authored blocks one
    // level closer to the root, preserving chart fields and table columns. Web
    // restores from that state as well, so prefer it when reopening on Android.
    val stateBlocks = windowForm.values.asSequence()
        .mapNotNull(::obj)
        .mapNotNull { it["reportDocumentBlocks"] as? JsonArray }
        .firstOrNull { it.isNotEmpty() }
    val stateDocument = stateBlocks?.let { blocks ->
        JsonObject((definitionDocument?.toMutableMap() ?: mutableMapOf()).apply {
            put("blocks", blocks)
        })
    }
    val candidates = listOf(
        stateDocument,
        reportDefinition?.get("documentPatch"),
        reportDefinition?.get("reportDocument"),
        windowForm["documentPatch"],
        windowForm["reportDocument"]
    )
    return candidates.asSequence()
        .mapNotNull(::obj)
        .firstOrNull { (it["blocks"] as? JsonArray)?.isNotEmpty() == true }
}

internal fun reportBuilderAuthoredDatasetRefs(document: JsonObject): Set<String> =
    (document["blocks"] as? JsonArray).orEmpty()
        .mapNotNull { block ->
            ((block as? JsonObject)?.get("datasetRef") as? JsonPrimitive)?.content
                ?.trim()?.takeIf { it.isNotEmpty() }
        }
        .toSet()

internal fun reportBuilderMaterializeComputedRows(
    rows: List<Map<String, Any?>>,
    config: DashboardReportBuilderDef
): List<Map<String, Any?>> {
    val computedKeys = config.computedMeasures.map(::reportBuilderMeasureKey)
    if (computedKeys.isEmpty()) return rows
    return rows.map { row -> applyReportBuilderComputedMeasures(row, computedKeys, config) }
}

internal fun reportBuilderPersistedDatasets(
    windowForm: Map<String, Any?>,
    config: DashboardReportBuilderDef
): Map<String, List<Map<String, Any?>>> {
    val datasets = windowForm["reportStaticDatasets"] as? List<*> ?: return emptyMap()
    return datasets.mapNotNull { value ->
        val dataset = value as? Map<*, *> ?: return@mapNotNull null
        val id = (dataset["id"] ?: dataset["dataSourceRef"])?.toString()?.trim().orEmpty()
        if (id.isEmpty()) return@mapNotNull null
        val rows = (dataset["rows"] as? List<*>).orEmpty().mapNotNull { row ->
            (row as? Map<*, *>)?.entries?.associate { it.key.toString() to it.value }
        }
        id to reportBuilderMaterializeComputedRows(rows, config)
    }.toMap()
}

internal fun reportBuilderPublishedSources(
    config: DashboardReportBuilderDef,
    document: JsonObject
): List<ReportBuilderPublishedDataSourceDef> {
    val referencedInDocumentOrder = (document["blocks"] as? JsonArray).orEmpty()
        .mapNotNull { block ->
            ((block as? JsonObject)?.get("datasetRef") as? JsonPrimitive)?.content
                ?.trim()?.takeIf { it.isNotEmpty() && it != "primary" }
        }
        .distinct()
    val order = referencedInDocumentOrder.withIndex().associate { it.value to it.index }
    return config.dataSources
        .filter { it.id in order }
        .sortedWith(compareBy<ReportBuilderPublishedDataSourceDef>(
            { reportBuilderPublishedFetchPriority(it) },
            { order[it.id] ?: Int.MAX_VALUE }
        ))
}

/** Fetch cheap aggregate cards before detailed chart/table datasets on mobile. */
internal fun reportBuilderPublishedFetchPriority(source: ReportBuilderPublishedDataSourceDef): Int {
    val request = source.request
    val dimensions = request["dimensions"] as? JsonObject
    val limit = (request["limit"] as? JsonPrimitive)?.content?.toIntOrNull()
    return if (dimensions?.isEmpty() == true && limit != null && limit <= 1) 0 else 1
}

/** Catalog request fields define the dataset shape; active filters remain inherited. */
internal fun reportBuilderPublishedRequest(
    primaryRequest: Map<String, Any?>,
    declaration: ReportBuilderPublishedDataSourceDef
): Map<String, Any?> {
    val catalog = JsonUtil.elementToAny(declaration.request) as? Map<*, *> ?: emptyMap<Any?, Any?>()
    val result = primaryRequest.toMutableMap()
    catalog.forEach { (rawKey, value) ->
        val key = rawKey?.toString() ?: return@forEach
        if (key == "filters") {
            val inherited = (result[key] as? Map<*, *>)?.entries
                ?.associate { it.key.toString() to it.value }.orEmpty()
            val declared = (value as? Map<*, *>)?.entries
                ?.associate { it.key.toString() to it.value }.orEmpty()
            result[key] = inherited + declared
        } else {
            result[key] = value
        }
    }
    return result
}

internal fun materializeReportBuilderAuthoredDocument(document: JsonObject): JsonObject {
    val blocks = (document["blocks"] as? JsonArray).orEmpty().map { value ->
        val block = value as? JsonObject ?: return@map value
        if ((block["kind"] as? JsonPrimitive)?.content != "chartBlock" || block["chartModel"] != null) {
            return@map block
        }
        val spec = block["chartSpec"] as? JsonObject ?: return@map block
        val xField = (spec["xField"] as? JsonPrimitive)?.content ?: return@map block
        val yFields = (spec["yFields"] as? JsonArray).orEmpty()
            .mapNotNull { (it as? JsonPrimitive)?.content }
        if (yFields.isEmpty()) return@map block
        val authoredType = (spec["type"] as? JsonPrimitive)?.content?.lowercase().orEmpty()
        val type = when (authoredType) {
            "horizontal_bar", "horizontalbar", "column" -> "bar"
            "stackedbar" -> "stacked_bar"
            else -> authoredType.ifBlank { "line" }
        }
        val chart = ChartDef(
            title = (spec["title"] as? JsonPrimitive)?.content
                ?: (block["title"] as? JsonPrimitive)?.content,
            xAxis = ChartAxisDef(dataKey = xField),
            yAxis = ChartAxisDef(label = (spec["yLabel"] as? JsonPrimitive)?.content),
            series = ChartSeriesDef(values = yFields.map { field ->
                ChartValueOption(name = field, label = field, value = field)
            }),
            type = type
        )
        JsonObject(block.toMutableMap().apply {
            put("chartModel", JsonUtil.json.parseToJsonElement(JsonUtil.json.encodeToString(chart)))
        })
    }
    return JsonObject(document.toMutableMap().apply { put("blocks", JsonArray(blocks)) })
}

@Composable
internal fun ReportBuilderAuthoredResult(
    runtime: ForgeRuntime,
    window: WindowContext,
    dashboardRoot: ContainerDef,
    config: DashboardReportBuilderDef,
    document: JsonObject,
    primaryRows: List<Map<String, Any?>>,
    primaryControl: ControlState,
    primaryRequest: Map<String, Any?>,
    runRequestId: String?
) {
    val referencedDatasetRefs = remember(document) { reportBuilderAuthoredDatasetRefs(document) }
    val persistedRows = reportBuilderPersistedDatasets(window.peekWindowForm(), config)
    val declarations = remember(config, document) { reportBuilderPublishedSources(config, document) }
    val contexts = declarations.mapNotNull { declaration ->
        window.contextForInstanceOrNull(
            instanceRef = "reportDocument:${declaration.id}",
            dataSourceRef = declaration.dataSourceRef
        )?.let { declaration to it }
    }
    val rowsById = linkedMapOf<String, List<Map<String, Any?>>>()
    rowsById.putAll(persistedRows)
    val controls = mutableListOf<ControlState>()
    if ("primary" in referencedDatasetRefs && "primary" !in persistedRows) {
        rowsById["primary"] = primaryRows
        controls += primaryControl
    }
    contexts.forEach { (declaration, context) ->
        if (declaration.id in persistedRows) return@forEach
        val rows by context.collection.flow.collectAsState(initial = context.collection.peek())
        val control by context.control.flow.collectAsState(initial = context.control.peek())
        rowsById[declaration.id] = reportBuilderMaterializeComputedRows(rows, config)
        controls += control
    }
    // A large authored report can reference many logical datasets backed by
    // one expensive cube. Hydrate them in priority order so Android does not
    // stampede the gateway with identical concurrent cube jobs. The first
    // cheap aggregate normally unlocks the KPI overview immediately.
    LaunchedEffect(primaryRequest, declarations, primaryRows, runRequestId) {
        val requestId = runRequestId?.trim()?.takeIf(String::isNotEmpty)
            ?: "native-${java.util.UUID.randomUUID()}"
        if (runRequestId.isNullOrBlank() && persistedRows.isNotEmpty()) {
            runtime.publishNativeReportMaterialization(
                windowId = window.windowId,
                requestId = requestId,
                status = "completed",
                rowsById = persistedRows,
                errors = emptyList()
            )
            return@LaunchedEffect
        }
        runtime.publishNativeReportMaterialization(
            windowId = window.windowId,
            requestId = requestId,
            status = "running",
            rowsById = emptyMap(),
            errors = emptyList()
        )
        val loadedRows = linkedMapOf<String, List<Map<String, Any?>>>()
        val loadErrors = mutableListOf<String>()
        if ("primary" in referencedDatasetRefs) {
            loadedRows["primary"] = primaryRows
            primaryControl.error?.takeIf(String::isNotBlank)?.let(loadErrors::add)
        }
        contexts.forEach { (declaration, context) ->
            val request = reportBuilderPublishedRequest(primaryRequest, declaration)
            if (request.isEmpty()) return@forEach
            context.setInputParameters(request, fetch = true)
            context.control.flow.first { it.loading || it.resolved }
            context.control.flow.first { !it.loading && it.resolved }
            loadedRows[declaration.id] = reportBuilderMaterializeComputedRows(
                context.collection.peek(),
                config
            )
            context.control.peek().error?.takeIf(String::isNotBlank)?.let(loadErrors::add)
        }
        runtime.publishNativeReportMaterialization(
            windowId = window.windowId,
            requestId = requestId,
            status = if (loadErrors.isEmpty()) "completed" else "failed",
            rowsById = loadedRows,
            errors = loadErrors
        )
    }

    val preparedDocument = remember(document) { materializeReportBuilderAuthoredDocument(document) }
    val artifact = remember(preparedDocument, rowsById.toMap()) {
        InlineReportRuntimeCompiler.compile(
            TranscriptCanonicalReport(
                scope = "report-builder",
                id = "${window.windowId}-authored-report",
                grammar = "report-document-v1",
                status = "ready",
                source = preparedDocument,
                dataSources = rowsById.mapValues { (id, rows) ->
                    TranscriptCanonicalData(
                        id = id,
                        format = "json",
                        payload = JsonArray(rows.map(JsonUtil::anyToElement))
                    )
                }
            )
        )
    }
    val runtimeContainer = remember(artifact.metadata) {
        artifact.metadata.view?.content?.containers?.firstOrNull()
    }
    val pending = controls.any { it.loading || !it.resolved }
    val hasMaterializedRows = rowsById.values.any { it.isNotEmpty() }
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        val error = controls.firstNotNullOfOrNull { it.error?.takeIf(String::isNotBlank) }
        if (error != null) {
            Text(
                text = authoredReportLoadErrorMessage(error),
                style = MaterialTheme.typography.bodySmall,
                color = Color(0xFFB42318)
            )
        }
        if (pending) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                CircularProgressIndicator(modifier = Modifier.size(18.dp), strokeWidth = 2.dp)
                Text(
                    text = "Loading report data…",
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
        // Do not turn an unresolved dataset into a definitive empty KPI/card.
        // Partial output remains useful after an explicit error, while a normal
        // in-flight request gets one honest loading state.
        if (runtimeContainer != null && (hasMaterializedRows || !pending || error != null)) {
            DashboardReportRuntimeSurface(runtime, window, runtimeContainer, dashboardRoot)
        }
    }
}

private fun ForgeRuntime.publishNativeReportMaterialization(
    windowId: String,
    requestId: String,
    status: String,
    rowsById: Map<String, List<Map<String, Any?>>>,
    errors: List<String>
) {
    val materialization = linkedMapOf<String, Any?>(
        "id" to requestId,
        "requestId" to requestId,
        "status" to status,
        "materialized" to (status == "completed"),
        "datasetRefs" to rowsById.keys.sorted(),
        "rowCounts" to rowsById.mapValues { it.value.size }
    )
    if (errors.isNotEmpty()) materialization["errors"] = errors
    val values = linkedMapOf<String, Any?>("reportMaterialization" to materialization)
    if (status != "running") {
        values["reportStaticDatasets"] = rowsById.keys.sorted().map { id ->
            mapOf(
                "id" to id,
                "dataSourceRef" to id,
                "rows" to rowsById[id].orEmpty()
            )
        }
    }
    setWindowFormValues(windowId, values, replace = false, bumpPrefillRevision = false)
}

internal fun authoredReportLoadErrorMessage(error: String): String {
    val detail = error.trim()
    if (detail.contains("504") || detail.contains("gateway time-out", ignoreCase = true)) {
        return "Report data took too long to load. Try refreshing."
    }
    if (detail.contains("timeout", ignoreCase = true) ||
        detail.contains("timed out", ignoreCase = true)
    ) {
        return "Some report data did not respond. Try refreshing."
    }
    return "Some report data could not be loaded. Try refreshing."
}

File: ios/Sources/ForgeIOSUI/ReportBuilderAuthoredRuntime.swift
import SwiftUI
import ForgeIOSRuntime

func reportBuilderAuthoredDocument(_ windowForm: [String: JSONValue]) -> [String: JSONValue]? {
    let definition = windowForm["reportDefinition"]?.objectValue
    let definitionDocument = definition?["documentPatch"]?.objectValue
        ?? definition?["reportDocument"]?.objectValue
    let stateContainers = windowForm.values.compactMap(\.objectValue)
    let stateBlocks = stateContainers.compactMap {
        $0["reportDocumentBlocks"]?.arrayValue
    }.first { !$0.isEmpty }
    let nestedDocument = stateContainers.compactMap {
        $0["reportDocument"]?.objectValue ?? $0["documentPatch"]?.objectValue
    }.first
    var stateDocument = nestedDocument ?? definitionDocument
    if let stateBlocks {
        var next = stateDocument ?? [:]
        next["blocks"] = .array(stateBlocks)
        stateDocument = next
    }
    let candidates = [
        stateDocument,
        definition?["documentPatch"]?.objectValue,
        definition?["reportDocument"]?.objectValue,
        windowForm["documentPatch"]?.objectValue,
        windowForm["reportDocument"]?.objectValue
    ]
    return candidates.compactMap { $0 }.first { !($0["blocks"]?.arrayValue ?? []).isEmpty }
}

func reportBuilderAuthoredDatasetRefs(_ document: [String: JSONValue]) -> [String] {
    (document["blocks"]?.arrayValue ?? []).compactMap {
        nonBlankAuthored($0.objectValue?["datasetRef"]?.stringValue)
    }
}

func reportBuilderPublishedSources(
    config: DashboardReportBuilderDef,
    document: [String: JSONValue]
) -> [ReportBuilderPublishedDataSourceDef] {
    let ordered = reportBuilderAuthoredDatasetRefs(document).filter { $0 != "primary" }
    let order = ordered.enumerated().reduce(into: [String: Int]()) { result, entry in
        if result[entry.element] == nil {
            result[entry.element] = entry.offset
        }
    }
    var seen = Set<String>()
    return config.dataSources
        .filter { order[$0.id] != nil }
        .filter { seen.insert($0.id).inserted }
        .sorted {
            let left = reportBuilderPublishedFetchPriority($0)
            let right = reportBuilderPublishedFetchPriority($1)
            return left == right ? (order[$0.id] ?? .max) < (order[$1.id] ?? .max) : left < right
        }
}

func reportBuilderPublishedFetchPriority(_ source: ReportBuilderPublishedDataSourceDef) -> Int {
    let dimensions = source.request["dimensions"]?.objectValue
        ?? Dictionary(uniqueKeysWithValues: source.fields.compactMap { field in
            guard field["kind"]?.stringValue == "dimension",
                  let key = nonBlankAuthored(field["key"]?.stringValue) else { return nil }
            return (key, .bool(true))
        })
    let limit: Int?
    switch source.request["limit"] {
    case .number(let value): limit = Int(value)
    case .string(let value): limit = Int(value)
    default: limit = nil
    }
    guard dimensions.isEmpty else { return 1 }
    if let limit {
        return limit <= 1 ? 0 : 1
    }
    // Hosted report payloads may publish the field catalog while deliberately
    // omitting the writable request. A field-only dataset with no dimensions
    // is still an aggregate/KPI request and should populate before charts and
    // detail tables.
    return source.fields.isEmpty ? 1 : 0
}

func reportBuilderPublishedRequest(
    primaryRequest: [String: JSONValue],
    declaration: ReportBuilderPublishedDataSourceDef,
    now: Date = Date(),
    calendar: Calendar = .current
) -> [String: JSONValue] {
    var inherited: [String: JSONValue] = [:]
    for key in ["filters", "refinements", "timeoutMs"] where primaryRequest[key] != nil {
        inherited[key] = primaryRequest[key]
    }
    let generated = reportBuilderPublishedFieldRequest(
        fields: declaration.fields,
        primaryRequest: primaryRequest
    )
    var result = mergeAuthoredObjects(inherited, mergeAuthoredObjects(generated, declaration.request))
    guard let relative = declaration.scope["relativeDateRange"]?.objectValue,
          let range = reportBuilderRelativeDateRange(
            preset: relative["preset"]?.stringValue,
            now: now,
            calendar: calendar
          ) else {
        return result
    }
    if let path = nonBlankAuthored(relative["startParamPath"]?.stringValue) {
        setAuthoredNestedValue(&result, path: path, value: .string(range.start))
    }
    if let path = nonBlankAuthored(relative["endParamPath"]?.stringValue) {
        setAuthoredNestedValue(&result, path: path, value: .string(range.end))
    }
    return result
}

private func reportBuilderPublishedFieldRequest(
    fields: [[String: JSONValue]],
    primaryRequest: [String: JSONValue]
) -> [String: JSONValue] {
    guard !fields.isEmpty else { return [:] }
    let dimensions: [String: JSONValue] = Dictionary(uniqueKeysWithValues: fields.compactMap { field -> (String, JSONValue)? in
        guard field["kind"]?.stringValue == "dimension",
              let key = nonBlankAuthored(field["key"]?.stringValue) else { return nil }
        return (key, .bool(true))
    })
    let measures: [String: JSONValue] = Dictionary(uniqueKeysWithValues: fields.compactMap { field -> (String, JSONValue)? in
        guard field["kind"]?.stringValue == "measure",
              let key = nonBlankAuthored(field["key"]?.stringValue) else { return nil }
        return (key, .bool(true))
    })
    var result: [String: JSONValue] = [
        "dimensions": .object(dimensions),
        "measures": .object(measures),
        "offset": .number(0),
        "limit": dimensions.isEmpty ? .number(1) : (primaryRequest["limit"] ?? .number(100))
    ]
    if let firstDimension = dimensions.keys.first {
        let orderField = measures["totalSpend"] != nil ? "totalSpend desc" : "\(firstDimension) asc"
        result["orderBy"] = .array([.string(orderField)])
    }
    return result
}

func reportBuilderMaterializeComputedRows(
    _ rows: [[String: JSONValue]],
    fields: [[String: JSONValue]],
    config: DashboardReportBuilderDef
) -> [[String: JSONValue]] {
    let computed = Set(fields.compactMap { field -> String? in
        guard field["kind"]?.stringValue == "computedMeasure" else { return nil }
        return nonBlankAuthored(field["key"]?.stringValue)
    })
    guard !computed.isEmpty else { return rows }
    return rows.map { row in
        var result = row
        for measure in config.computedMeasures where computed.contains(measure.identityKey) {
            guard result[measure.identityKey] == nil,
                  let compute = measure.compute,
                  (compute.type ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased() == "ratio",
                  let numerator = compute.numerator,
                  let denominator = compute.denominator,
                  let denominatorValue = authoredNumber(result[denominator]),
                  denominatorValue != 0 else { continue }
            var value = (authoredNumber(result[numerator]) ?? 0) / denominatorValue * (compute.scale ?? 1)
            if let decimals = compute.decimals {
                let factor = pow(10.0, Double(max(0, decimals)))
                value = (value * factor).rounded() / factor
            }
            result[measure.identityKey] = .number(value)
        }
        return result
    }
}

private func authoredNumber(_ value: JSONValue?) -> Double? {
    switch value {
    case .number(let number): return number
    case .string(let string): return Double(string)
    default: return nil
    }
}

func materializeReportBuilderAuthoredDocument(_ document: [String: JSONValue]) -> [String: JSONValue] {
    var result = document
    result["blocks"] = .array((document["blocks"]?.arrayValue ?? []).map { value in
        guard var block = value.objectValue,
              block["kind"]?.stringValue == "chartBlock",
              block["chartModel"] == nil,
              let spec = block["chartSpec"]?.objectValue,
              let xField = nonBlankAuthored(spec["xField"]?.stringValue) else {
            return value
        }
        let yFields = (spec["yFields"]?.arrayValue ?? []).compactMap { nonBlankAuthored($0.stringValue) }
        guard !yFields.isEmpty else { return value }
        let authoredType = spec["type"]?.stringValue?.lowercased() ?? ""
        let type: String
        switch authoredType {
        case "horizontal_bar", "horizontalbar", "column": type = "bar"
        case "stackedbar": type = "stacked_bar"
        default: type = authoredType.isEmpty ? "line" : authoredType
        }
        block["chartModel"] = .object([
            "title": spec["title"] ?? block["title"] ?? .null,
            "type": .string(type),
            "xAxis": .object(["dataKey": .string(xField)]),
            "yAxis": .object(["label": spec["yLabel"] ?? .null]),
            "series": .object(["values": .array(yFields.map { field in
                .object(["name": .string(field), "value": .string(field)])
            })])
        ])
        return .object(block)
    })
    return result
}

func authoredReportLoadErrorMessage(_ error: String) -> String {
    let detail = error.lowercased()
    if detail.contains("504") || detail.contains("gateway time-out") {
        return "Report data took too long to load. Try refreshing."
    }
    if detail.contains("timeout") || detail.contains("timed out") {
        return "Some report data did not respond. Try refreshing."
    }
    return "Some report data could not be loaded. Try refreshing."
}

struct ReportBuilderAuthoredResult: View {
    let runtime: ForgeRuntime
    let window: WindowContext
    let dashboardRoot: ContainerDef
    let config: DashboardReportBuilderDef
    let document: [String: JSONValue]
    let primaryRows: [[String: JSONValue]]
    let primaryControl: ControlState
    let primaryRequest: [String: JSONValue]
    let runRequestID: String?

    @State private var rowsByID: [String: [[String: JSONValue]]] = [:]
    @State private var controlsByID: [String: ControlState] = [:]

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            if let error = visibleError {
                Label(authoredReportLoadErrorMessage(error), systemImage: "exclamationmark.triangle")
                    .font(.footnote)
                    .foregroundStyle(.red)
            }
            if isLoading {
                HStack(spacing: 8) {
                    ProgressView().controlSize(.small)
                    Text("Loading report data…").font(.subheadline).foregroundStyle(.secondary)
                }
                .accessibilityIdentifier("forge-authored-report-loading")
            }
            if let container = runtimeContainer,
               hasMaterializedRows || !isLoading || visibleError != nil {
                DashboardRenderer(runtime: runtime, window: window, container: container)
            }
        }
        .task(id: materializationTaskID) {
            await loadPublishedDatasets()
        }
    }

    private var declarations: [ReportBuilderPublishedDataSourceDef] {
        reportBuilderPublishedSources(config: config, document: document)
    }

    private var allRows: [String: [[String: JSONValue]]] {
        var result = rowsByID
        if reportBuilderAuthoredDatasetRefs(document).contains("primary") {
            result["primary"] = primaryRows
        }
        return result
    }

    private var allControls: [ControlState] {
        var result = Array(controlsByID.values)
        if reportBuilderAuthoredDatasetRefs(document).contains("primary") {
            result.append(primaryControl)
        }
        return result
    }

    private var isLoading: Bool { allControls.contains { $0.loading } }
    private var hasMaterializedRows: Bool { allRows.values.contains { !$0.isEmpty } }
    private var visibleError: String? { allControls.compactMap(\.error).first { !$0.isEmpty } }

    private var runtimeContainer: ContainerDef? {
        let prepared = materializeReportBuilderAuthoredDocument(document)
        let sources = allRows.mapValues { rows in
            rows.map(JSONValue.object)
        }
        guard let artifact = try? InlineReportRuntimeCompiler.compile(
            TranscriptCanonicalReport(
                scope: "report-builder",
                id: "\(window.windowID)-authored-report",
                grammar: "report-document-v1",
                status: "ready",
                source: .object(prepared),
                dataSources: Dictionary(uniqueKeysWithValues: sources.map { id, rows in
                    (id, TranscriptCanonicalData(id: id, format: "json", payload: .array(rows)))
                })
            )
        ) else { return nil }
        return artifact.metadata.view?.content?.containers.first
    }

    private var requestSignature: String {
        let declarationSignature = declarations.map { "\($0.id):\($0.request.jsonSignature):\($0.scope.jsonSignature)" }.joined(separator: "|")
        return primaryRequest.jsonSignature + "::" + declarationSignature
    }

    private var materializationTaskID: String {
        requestSignature + "::" + (nonBlankAuthored(runRequestID) ?? "auto")
    }

    @MainActor
    private func loadPublishedDatasets() async {
        let requestID = nonBlankAuthored(runRequestID) ?? "native-\(UUID().uuidString)"
        await publishMaterialization(
            requestID: requestID,
            status: "running",
            rowsByID: [:],
            errors: []
        )
        rowsByID = [:]
        controlsByID = [:]
        var loadedRows: [String: [[String: JSONValue]]] = [:]
        var loadErrors: [String] = []
        if reportBuilderAuthoredDatasetRefs(document).contains("primary") {
            loadedRows["primary"] = primaryRows
            if let error = primaryControl.error, !error.isEmpty {
                loadErrors.append(error)
            }
        }
        for declaration in declarations {
            guard !Task.isCancelled else { return }
            controlsByID[declaration.id] = ControlState(loading: true)
            let request = reportBuilderPublishedRequest(primaryRequest: primaryRequest, declaration: declaration)
            let instanceRef = "reportDocument:\(declaration.id)"
            await runtime.fetchDataSourceInstance(
                windowID: window.windowID,
                instanceRef: instanceRef,
                dataSourceRef: declaration.dataSourceRef,
                parameters: request
            )
            guard !Task.isCancelled else { return }
            let rows = await runtime.dataSourceCollection(
                windowID: window.windowID,
                dataSourceRef: instanceRef
            )
            let materializedRows = reportBuilderMaterializeComputedRows(
                rows,
                fields: declaration.fields,
                config: config
            )
            rowsByID[declaration.id] = materializedRows
            loadedRows[declaration.id] = materializedRows
            let control = await runtime.dataSourceControl(
                windowID: window.windowID,
                dataSourceRef: instanceRef
            )
            controlsByID[declaration.id] = control
            if let error = control.error, !error.isEmpty {
                loadErrors.append(error)
            }
        }
        await publishMaterialization(
            requestID: requestID,
            status: loadErrors.isEmpty ? "completed" : "failed",
            rowsByID: loadedRows,
            errors: loadErrors
        )
    }

    @MainActor
    private func publishMaterialization(
        requestID: String,
        status: String,
        rowsByID: [String: [[String: JSONValue]]],
        errors: [String]
    ) async {
        let datasets = rowsByID.keys.sorted().map { id in
            JSONValue.object([
                "id": .string(id),
                "dataSourceRef": .string(id),
                "rows": .array((rowsByID[id] ?? []).map(JSONValue.object))
            ])
        }
        let rowCounts = Dictionary(uniqueKeysWithValues: rowsByID.map { id, rows in
            (id, JSONValue.number(Double(rows.count)))
        })
        var materialization: [String: JSONValue] = [
            "id": .string(requestID),
            "requestId": .string(requestID),
            "status": .string(status),
            "materialized": .bool(status == "completed"),
            "datasetRefs": .array(rowsByID.keys.sorted().map(JSONValue.string)),
            "rowCounts": .object(rowCounts)
        ]
        if !errors.isEmpty {
            materialization["errors"] = .array(errors.map(JSONValue.string))
        }
        var values: [String: JSONValue] = [
            "reportMaterialization": .object(materialization)
        ]
        if status != "running" {
            values["reportStaticDatasets"] = .array(datasets)
        }
        await runtime.setWindowFormValue(
            windowID: window.windowID,
            values: values,
            replace: false
        )
    }
}

private func reportBuilderRelativeDateRange(
    preset: String?,
    now: Date,
    calendar: Calendar
) -> (start: String, end: String)? {
    let key = (preset ?? "").lowercased().replacingOccurrences(of: "_", with: "")
    let today = calendar.startOfDay(for: now)
    let offsets: (Int, Int)?
    switch key {
    case "today": offsets = (0, 0)
    case "yesterday": offsets = (-1, -1)
    case "last3days", "3d": offsets = (-2, 0)
    case "last7days", "7d": offsets = (-6, 0)
    case "last30days", "30d": offsets = (-29, 0)
    default: offsets = nil
    }
    guard let offsets,
          let start = calendar.date(byAdding: .day, value: offsets.0, to: today),
          let end = calendar.date(byAdding: .day, value: offsets.1, to: today) else { return nil }
    let formatter = DateFormatter()
    formatter.calendar = calendar
    formatter.timeZone = calendar.timeZone
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.dateFormat = "yyyy-MM-dd"
    return (formatter.string(from: start), formatter.string(from: end))
}

private func mergeAuthoredObjects(
    _ inherited: [String: JSONValue],
    _ declared: [String: JSONValue]
) -> [String: JSONValue] {
    var result = inherited
    for (key, value) in declared {
        if let left = result[key]?.objectValue, let right = value.objectValue {
            result[key] = .object(mergeAuthoredObjects(left, right))
        } else {
            result[key] = value
        }
    }
    return result
}

private func setAuthoredNestedValue(_ object: inout [String: JSONValue], path: String, value: JSONValue) {
    let parts = path.split(separator: ".").map(String.init).filter { !$0.isEmpty }
    guard let head = parts.first else { return }
    if parts.count == 1 { object[head] = value; return }
    var child = object[head]?.objectValue ?? [:]
    setAuthoredNestedValue(&child, path: parts.dropFirst().joined(separator: "."), value: value)
    object[head] = .object(child)
}

private func nonBlankAuthored(_ value: String?) -> String? {
    let result = value?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    return result.isEmpty ? nil : result
}

private extension Dictionary where Key == String, Value == JSONValue {
    var jsonSignature: String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        guard let data = try? encoder.encode(self) else { return "" }
        return String(decoding: data, as: UTF8.self)
    }
}
